src/util/mpsc.rs

```diff
@@ -169,6 +190,27 @@ pub fn bounded<T>(cap: usize) -> (Sender<T>, Receiver<T>) {
     )
 }
 
+impl<T> Stream for Receiver<T> {
+    type Error = ();
+    type Item = T;
+
+    fn poll(&mut self) -> Poll<Option<T>, ()> {
+        match self.try_recv() {
+            Ok(m) => Ok(Async::Ready(Some(m))),
+            Err(TryRecvError::Empty) => {
+                let t = self.state.recv_task.swap(task::current(), Ordering::AcqRel);
+                if t.is_some() {
+                    return Ok(Async::NotReady);
+                }
+                // Note: If there is a message or all the senders are dropped after
+                // polling but before task is set, `t` should be None.
+                self.poll()
```

Comment:
Why not replace the recursion with a loop?

Reply:
It's easy to understand and there is only one level of recursion.